### ISSUE TITLE
Async / Await hanging issue with Soap requests

### ISSUE THREAD

**Issue by heskethw** [ServiceStack/Issues#640]
Hi,

I have noticed a bug since upgrading from version 4.5.14 with how Async / Await is handled in the .NET Framework hosted version, I have not tested in .NET Core, but have confirmed that reverting to version 4.5.14 resolves the issue. 

When a service interface method awaits an asynchronous task using the await keyword, **and** if you are using a SOAP client to connect to the interface, the SOAP request times out without a response. The result of the await doesn't appear to be returned to the main thread and the request hangs on the server. Using ConfigureAwait(false) does seem to allow the awaited task result to be returned to the thread and it steps through correctly in the debugger, but the response to the SOAP request still isn't returned to the client. 

I noticed the problem since our public facing API calls our internal JSON API and the requests via SOAP were failing when JSON and XML worked correctly. I haven't tested with a generic awaitable method, only with awaiting a URL/JSON request as in the example below.

The example I used to reproduce the bug used a placeholder JSON API and the following code:

**Service Interface**
```
public async Task<object> Any(Hello request)
{
  var url = $"https://jsonplaceholder.typicode.com/users/{request.Id}";

  var response = await url.GetJsonFromUrlAsync().ConfigureAwait(false);

  return new HelloResponse { Result = response };
}
```

**Service Model**
```
[Route("/hello/{Id}")]
public class Hello : IReturn<HelloResponse>
{
  public int Id { get; set; }
}

public class HelloResponse
{
  public string Result { get; set; }
}
```

**Service Tests**
```
[Test]
public void TestSoap12()
{
  var client = new Soap12ServiceClient("http://localhost:6982/");
	
  var response = client.Send<HelloResponse>(new Hello { Id = 1 });

  Assert.That(response.Result, Contains.Substring("Leanne Graham"));
}

[Test]
public void TestJson()
{
  var client = new JsonServiceClient("http://localhost:6982/");
	
  var response = client.Send<HelloResponse>(new Hello { Id = 1 });

  Assert.That(response.Result, Contains.Substring("Leanne Graham"));
}
```

I have reproduced the bug in a fresh copy of the project template [available here](https://github.com/heskethw/ServiceStackSoapTest) There are two tests, one for a SOAP client, one for a JSON client, both call the same API method, but only the JSON test succeeds.

I don't know if this is related to the web service awaiting a request to another web service, or awaiting tasks in general, but it's odd it only happens on SOAP requests. 

Let me know if you need anything else 👍 thanks for your time!

**Comment by mythz** [comment]
From v5 we've moved to use async/await internally and it looks like calling async Services from SOAP are causing deadlocks. I'm exploring to see if there's any workarounds that can avoid the deadlocks but so far I would avoid calling Async Services from SOAP.

**Comment by heskethw** [comment]
Thanks for looking into it, if there isn't an easy fix, we'll have to avoid declaring the API interfaces as async as you mentioned. I'm hoping it's a simple fix though!

**Comment by mythz** [comment]
Ok was able to refactor the SOAP Handlers to execute services using non-blocking async APIs which resolves the deadlocking issues from [this commit](https://github.com/ServiceStack/ServiceStack/commit/da8a3f9a604ec860ce4876f4fb5303e46e6710a7).

This change is available from v5.4.1 that's now [available on MyGet](http://docs.servicestack.net/myget).

**Comment by heskethw** [comment]
Hi Mythz. That's awesome, really appreciate you're quick fix and response, keep up the great work! 👍
